File: 수학_정수론/프로그래머스lv1-약수의 개수와 덧셈.py
def solution(left, right):
    #제곱수를 가지면 무조건 약수의 개수는 홀수임
    result = 0
    for i in range(left, right + 1):
        if int(i ** 0.5) == i ** 0.5:
            result -= i
        else:
            result += i
    
    return result

"""
정확성  테스트
테스트 1 〉	통과 (0.44ms, 10.3MB)
테스트 2 〉	통과 (0.21ms, 10.3MB)
테스트 3 〉	통과 (0.17ms, 10.3MB)
테스트 4 〉	통과 (0.03ms, 10.3MB)
테스트 5 〉	통과 (0.30ms, 10.3MB)
테스트 6 〉	통과 (0.05ms, 10.3MB)
테스트 7 〉	통과 (0.04ms, 10.3MB)
"""

# 2022년 7월
import math
def get_divisor(number):
    if number < 0:
        raise Exception("숫자는 0 이상의 정수여아합니다.")
    if number == 0:
        return [0]
    divisors = []
    limit = math.floor(math.sqrt(number))
    for left in range(1, limit + 1):
        if number % left == 0:
            right = number // left
            if left == right:
                divisors.append(left)
            else:
                divisors.append(left)
                divisors.append(right)
    return divisors

def solution(left, right):
    result = 0
    for number in range(left, right + 1):
        n = len(get_divisor(number))
        if n % 2 == 0:
            result += number
        else:
            result -= number
    return result
"""
정확성  테스트
테스트 1 〉	통과 (1.84ms, 10.3MB)
테스트 2 〉	통과 (0.50ms, 10.2MB)
테스트 3 〉	통과 (0.86ms, 10.2MB)
테스트 4 〉	통과 (0.12ms, 10.2MB)
테스트 5 〉	통과 (1.51ms, 10.1MB)
테스트 6 〉	통과 (0.14ms, 10.2MB)
테스트 7 〉	통과 (0.16ms, 10.2MB)
"""

File: 수학_정수론/프로그래머스lv1-약수의 개수와 덧셈.js
function solution(left, right) {
  function measure(number) {
    let nums = []
    const c = Math.sqrt(number)
    for (var i = 1; i < c; i++) {
      if (number % i == 0) {
        nums.push(i)
        nums.push(number / i)
      }
    }
    if (number % c == 0) {
      nums.push(c)
    }

    return nums.length
  }
  let result = 0
  for (var num = left; num <= right; num++) {
    let r = measure(num)
    r % 2 == 0 ? (result += num) : (result -= num)
  }
  return result
}
// 정확성  테스트
// 테스트 1 〉	통과 (2.21ms, 32.8MB)
// 테스트 2 〉	통과 (0.57ms, 30.1MB)
// 테스트 3 〉	통과 (0.66ms, 30.2MB)
// 테스트 4 〉	통과 (0.13ms, 29.9MB)
// 테스트 5 〉	통과 (1.74ms, 32.6MB)
// 테스트 6 〉	통과 (0.24ms, 29.9MB)
// 테스트 7 〉	통과 (0.18ms, 30MB)

const getDivisors = (number) => {
  if (number < 0) throw new Error('숫자는 0 이상의 정수여야합니다.')
  if (number === 0) return [0]
  const limit = Math.floor(Math.sqrt(number))
  const iters = Array.from({ length: limit }, (_, i) => i + 1)
  return iters.reduce((divisors, left) => {
    if (number % left === 0) {
      const right = number / left
      if (left === right) {
        divisors.push(left)
        return divisors
      }
      divisors.push(left)
      divisors.push(right)
      return divisors
    }
    return divisors
  }, [])
}

function solution(left, right) {
  let result = 0
  for (let number = left; number <= right; number++) {
    const n = getDivisors(number).length
    if (n % 2 === 0) {
      result += number
    } else {
      result -= number
    }
  }
  return result
}
// 정확성  테스트
// 테스트 1 〉	통과 (22.93ms, 33.2MB)
// 테스트 2 〉	통과 (1.41ms, 29.9MB)
// 테스트 3 〉	통과 (1.93ms, 30.2MB)
// 테스트 4 〉	통과 (0.58ms, 30.1MB)
// 테스트 5 〉	통과 (23.95ms, 33.3MB)
// 테스트 6 〉	통과 (0.66ms, 30.2MB)
// 테스트 7 〉	통과 (0.44ms, 30.1MB)
